mavros/src/plugins/waypoint.cpp

```diff
@@ -98,33 +172,110 @@ class WaypointItem : public mavlink::common::msg::MISSION_ITEM {
-		// [[[end]]] (checksum: e277c889ab7a67085562bbd014283a78)
+		// [[[end]]] (checksum: 3fb825422253905ba8e3256be09a6f1a)
 
 		ret.seq = seq;
 		ret.mission_type = enum_value(mavlink::common::MAV_MISSION_TYPE::MISSION);
 
 		return ret;
 	}
+};
+
+class WaypointItemInt : public mavlink::common::msg::MISSION_ITEM_INT {
+public:
+
+	mavros_msgs::Waypoint to_msg() const
+	{
+		mavros_msgs::Waypoint ret;
+
+		// [[[cog:
+		// for a, b in waypoint_item_msg + waypoint_coords:
+		//     if a[0] == 'x' or a[0] == 'y':
+		//         cog.outl("ret.%s = double(%s) / double(waypoint_encode_factor(frame));" % (a,b))
+		//     else:
+		//         cog.outl("ret.%s = %s;" % (a, b))
+		// ]]]
+		ret.frame = frame;
+		ret.command = command;
+		ret.is_current = current;
+		ret.autocontinue = autocontinue;
+		ret.param1 = param1;
+		ret.param2 = param2;
+		ret.param3 = param3;
+		ret.param4 = param4;
+		ret.x_lat = double(x) / double(waypoint_encode_factor(frame));
+		ret.y_long = double(y) / double(waypoint_encode_factor(frame));
+		ret.z_alt = z;
+		// [[[end]]] (checksum: 757604153b120847803e4bc4a8bccf9b)
+
+		return ret;
+	}
 
-	std::string to_string()
+	static WaypointItemInt from_msg(mavros_msgs::Waypoint &wp, uint16_t seq)
 	{
-		//return to_yaml();
+		WaypointItemInt ret{};
+
+		// [[[cog:
+		// for a, b in waypoint_item_msg + waypoint_coords:
+		//     if b[0] == 'x' or b[0] == 'y':
+		//         cog.outl("ret.%s = int32_t(wp.%s * waypoint_encode_factor(wp.frame));" % (b,a))
+		//     else:
+		//         cog.outl("ret.%s = wp.%s;" % (b,a))
+		// ]]]
+		ret.frame = wp.frame;
+		ret.command = wp.command;
+		ret.current = wp.is_current;
+		ret.autocontinue = wp.autocontinue;
+		ret.param1 = wp.param1;
+		ret.param2 = wp.param2;
+		ret.param3 = wp.param3;
+		ret.param4 = wp.param4;
+		ret.x = int32_t(wp.x_lat * waypoint_encode_factor(wp.frame));
+		ret.y = int32_t(wp.y_long * waypoint_encode_factor(wp.frame));
+		ret.z = wp.z_alt;
+		// [[[end]]] (checksum: bf26a63f03988e41aa372667edcae7d8)
+
+		ret.seq = seq;
+		ret.mission_type = enum_value(mavlink::common::MAV_MISSION_TYPE::MISSION);
+
+		return ret;
+	}
+};
 
-		return utils::format("#%u%1s F:%u C:%3u p: %f %f %f %f x: %f y: %f z: %f",
-			seq,
-			(current) ? "*" : "",
-			frame, command,
-			param1, param2, param3, param4,
-			x_lat, y_long, z_alt);
+struct to_msg_visitor : public boost::static_visitor<mavros_msgs::Waypoint>{
+public:
+	mavros_msgs::Waypoint operator()( const WaypointItemInt& itemInt ) const {
+		return itemInt.to_msg();
+	}
+
+	mavros_msgs::Waypoint operator()( const WaypointItem& item ) const {
+		return item.to_msg();
 	}
 };
+template <class T>
+std::string waypoint_to_string(const T &waypoint){
+	return utils::format("#%u%1s F:%u C:%3u p: %f %f %f %f x: %i y: %i z: %i",
+		waypoint.seq,
+		(waypoint.current) ? "*" : "",
+		waypoint.frame, waypoint.command,
+		waypoint.param1, waypoint.param2, waypoint.param3, waypoint.param4,
+		waypoint.x, waypoint.y, waypoint.z);
+}
```

Comment:
I think better to keep to_string in classes. You shouldn't use `%i` for float.
Also why not just to define common interface class for both variants?

Reply:
That might be better, so originally I had been planning on storing the variant type in the waypoint and send_waypoint vector and so I believe the lack of interface is left over from that because my first attempt was to give them a parent class and store that in the vector which didnt work very elegantly due to the difference in member variable types but I think I could give them a parent template class with this implementation which inherits from either MISSION_ITEM or MISSION_ITEM_INT and gives them a to_string in the class.

Reply:
I realized that since I'm not using the classes to store data as I was originally intending I can now just give them to from ros message template functions and it should have cleaned everything up.